File: examples/java/src/main/java/io/superdurable/dex/products/signup/UserSignupFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.products.signup;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import io.superdurable.dex.Timer;
import io.superdurable.dex.Wait;
import io.superdurable.dex.shared.MyDependencyService;
import org.springframework.stereotype.Component;

import java.time.Duration;

@Component
public class UserSignupFlow implements Flow<SignupForm> {
    public final Attribute<SignupForm> form = Attribute.define("Form", SignupForm.class);
    public final Attribute<String> status = Attribute.define("Status", String.class);
    public final Channel<Void> verify = Channel.define("Verify", Void.class);

    private final MyDependencyService service;
    private final Submit submit = new Submit();
    private final Verify verifyStep = new Verify();

    public UserSignupFlow(final MyDependencyService service) {
        this.service = service;
    }

    @Override
    public StepList<SignupForm> getSteps() {
        return StepList.startStep(submit).otherSteps(verifyStep);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(form, status, verify);
    }

    @RPC
    public RPCResult<String> verify(final Context context) {
        if ("verified".equals(status.get(context))) {
            return RPCResult.of("already verified");
        }
        status.set(context, "verified");
        verify.publish(context, null);
        return RPCResult.of("done");
    }

    final class Submit implements Step<SignupForm> {
        @Override
        public Class<SignupForm> getInputType() {
            return SignupForm.class;
        }

        @Override
        public StepDecision execute(final Context context, final SignupForm input) {
            form.set(context, input);
            status.set(context, "waiting");
            service.sendEmail(input.email, "please verify the signup", "content");
            return StepDecision.goTo(verifyStep, null);
        }
    }

    final class Verify implements Step<Void> {
        @Override
        public Class<Void> getInputType() {
            return Void.class;
        }

        @Override
        public Wait waitFor(final Context context, final Void input) {
            return Wait.anyOf(
                    Timer.byDuration(Duration.ofSeconds(24)),
                    verify.forOne());
        }

        @Override
        public StepDecision execute(final Context context, final Void input) {
            final SignupForm signupForm = form.get(context);
            if (!verify.getConditionResults(context).isEmpty()) {
                service.sendEmail(signupForm.email, "welcome", "welcome to Indeed!");
                return StepDecision.gracefulComplete("done");
            }
            service.sendEmail(signupForm.email, "reminder", "please verify your email");
            return StepDecision.goTo(verifyStep, null);
        }
    }
}

File: examples/python/dex_examples/products/signup/user_signup_flow.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from datetime import timedelta

from dex import (
    Attribute,
    Channel,
    Context,
    Flow,
    PersistenceSchema,
    RPCResult,
    Step,
    StepDecision,
    StepList,
    Timer,
    Wait,
    go_to,
    graceful_complete,
    rpc,
)

from dex_examples.shared.my_dependency_service import MyDependencyService
from dex_examples.products.signup.signup_form import SignupForm


class Verify(Step[None]):
    def __init__(
        self,
        service: MyDependencyService,
        form: Attribute[SignupForm],
        verify_channel: Channel[None],
    ) -> None:
        self.service = service
        self.form = form
        self.verify_channel = verify_channel

    def wait_for(self, context: Context, input: None) -> Wait:
        del context, input
        return Wait.any_of(
            Timer.by_duration(timedelta(seconds=24)),
            self.verify_channel.for_one(),
        )

    def execute(self, context: Context, input: None) -> StepDecision:
        del input
        signup_form = self.form.get(context)
        if self.verify_channel.results(context):
            self.service.send_email(signup_form.email, "welcome", "welcome to Indeed!")
            return graceful_complete("done")
        self.service.send_email(
            signup_form.email,
            "reminder",
            "please verify your email",
        )
        return go_to(Verify, None)


class Submit(Step[SignupForm]):
    def __init__(
        self,
        service: MyDependencyService,
        form: Attribute[SignupForm],
        status: Attribute[str],
        verify_step: Verify,
    ) -> None:
        self.service = service
        self.form = form
        self.status = status
        self.verify_step = verify_step

    def execute(self, context: Context, input: SignupForm) -> StepDecision:
        self.form.set(context, input)
        self.status.set(context, "waiting")
        self.service.send_email(input.email, "please verify the signup", "content")
        return go_to(Verify, None)


class UserSignupFlow(Flow[SignupForm]):
    form = Attribute("Form", SignupForm)
    status = Attribute("Status", str)
    verify_channel = Channel[None]("Verify", type(None))

    def __init__(self, service: MyDependencyService) -> None:
        self.service = service
        self.verify_step = Verify(service, self.form, self.verify_channel)
        self.submit = Submit(service, self.form, self.status, self.verify_step)

    def get_steps(self) -> StepList[SignupForm]:
        return StepList.start_step(self.submit).other_steps(self.verify_step)

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(self.form, self.status, self.verify_channel)

    @rpc
    def verify(self, context: Context) -> RPCResult[str]:
        if self.status.get(context) == "verified":
            return RPCResult("already verified")
        self.status.set(context, "verified")
        self.verify_channel.publish(context, None)
        return RPCResult("done")
